### PR TITLE
add staging flag for hitting staging servers

### PR DESCRIPTION
**Pull request by myleshorton**

### CODE REVIEW

File: src/github.com/getlantern/flashlight/config/config.go [hunk 1]
```diff
@@ -59,9 +59,9 @@ type Fetcher interface {
 }
 
 // StartPolling starts the process of polling for new configuration files.
-func StartPolling() {
+func (cfg *Config) StartPolling() {
 	// Force detour to whitelist chained domain
-	u, err := url.Parse(chainedCloudConfigURL)
+	u, err := url.Parse(cfg.CloudConfig)
```
Comment (myleshorton) [inline src/github.com/getlantern/flashlight/config/config.go]: We should really be using the configured server here rather than the default -- made this a method of the struct accordingly.

File: src/github.com/getlantern/flashlight/config/config.go [hunk 2]
```diff
@@ -78,6 +78,12 @@ func validateConfig(_cfg yamlconf.Config) error {
 	if !ok {
 		return fmt.Errorf("Config is not a flashlight config!")
 	}
+	if cfg.Client == nil {
+		return fmt.Errorf("No client config")
+	}
+	if cfg.Client.ChainedServers == nil {
+		return fmt.Errorf("No chained servers")
+	}
```
Comment (myleshorton) [inline src/github.com/getlantern/flashlight/config/config.go]: This actually just knocks out https://github.com/getlantern/lantern/issues/4357 while we're in here.

File: src/github.com/getlantern/flashlight/config/fetcher.go [hunk 3]
```diff
@@ -16,16 +16,17 @@ import (
...
-	ifNoneMatch           = "X-Lantern-If-None-Match"
-	userIDHeader          = "X-Lantern-User-Id"
-	tokenHeader           = "X-Lantern-Pro-Token"
-	chainedCloudConfigURL = "http://config.getiantem.org/cloud.yaml.gz"
+	etag         = "X-Lantern-Etag"
+	ifNoneMatch  = "X-Lantern-If-None-Match"
+	userIDHeader = "X-Lantern-User-Id"
+	tokenHeader  = "X-Lantern-Pro-Token"
+
+	defaultChainedCloudConfigURL = "http://config.getiantem.org/cloud.yaml.gz"
```
Comment (myleshorton) [inline src/github.com/getlantern/flashlight/config/fetcher.go]: Just a slightly clearer, if longer, name.

File: src/github.com/getlantern/flashlight/flashlight.go [hunk 4]
```diff
@@ -134,7 +134,7 @@ func Run(httpProxyAddr string,
 			// proxying capabilities of Lantern, so it needs everything to be properly
 			// set up with at least an initial bootstrap config (on first run) to
 			// complete successfully.
-			config.StartPolling()
+			cfg.StartPolling()
```
Comment (myleshorton) [inline src/github.com/getlantern/flashlight/flashlight.go]: Start polling on the actual config instance.

**Comment by myleshorton** [comment]
hey @oxtoacart you mind taking a look at this one?

File: src/github.com/getlantern/flashlight/config/config.go (hunk 2 above)
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/config/config.go]: Shouldn't this have been taken care of by ApplyDefaults()?

Comment (myleshorton) [inline src/github.com/getlantern/flashlight/config/config.go]: Yeah just removed it @oxtoacart -- I think that client-reported error isn't possible with what's on devel actually!

File: src/github.com/getlantern/flashlight/config/fetcher.go [hunk 5]
```diff
@@ -49,8 +57,43 @@ type UserConfig interface {
...
 // NewFetcher creates a new configuration fetcher with the specified
 // interface for obtaining the user ID and token if those are populated.
-func NewFetcher(conf UserConfig, rt http.RoundTripper) Fetcher {
-	return &fetcher{lastCloudConfigETag: map[string]string{}, user: conf, rt: rt}
+func NewFetcher(conf UserConfig, rt http.RoundTripper, flags map[string]interface{}) Fetcher {
+	var stage bool
+	if s, ok := flags["staging"].(bool); ok {
+		stage = s
+	} else {
+		stage = false
```
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/config/fetcher.go]: This is redundant, as stage is already false.

Comment (myleshorton) [inline src/github.com/getlantern/flashlight/config/fetcher.go]: Good catch should be all set!

File: src/github.com/getlantern/flashlight/config/config.go [hunk 6]
```diff
@@ -40,31 +40,25 @@ var (
...
+	MemProfile      string
+	UpdateServerURL string
+	Client          *client.ClientConfig
+	ProxiedSites    *proxiedsites.Config // List of proxied site domains that get routed through Lantern rather than accessed directly
+	TrustedCAs      []*fronted.CA
 }
 
 // StartPolling starts the process of polling for new configuration files.
-func StartPolling() {
+func (cfg *Config) StartPolling() {
```
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/config/config.go]: Why did you change polling to refer to a specific Config instance?  Polling is a process that continues forever, even though the underlying config instance changes over time.  Also, there's nothing in here that specifically refers to the cfg instance, so there's no benefit to it being an instance method.  Plus it refers to the global yamlconf.Manager.

File: src/github.com/getlantern/flashlight/flashlight.go (hunk 4 above)
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/flashlight.go]: See comment above - I would revert this to polling on config package.

**Comment by oxtoacart** [comment]
@myleshorton This now works when I test.  Travis test failed for what seems to be an unrelated issue, I've restarted it.

The basic logic looks solid to me, there's just a couple of things:

1. Polling on the Config instance doesn't make any sense
2. There's a redundant else branch assigning false to an already false boolean

Other than those and assuming the test passes this time, this LGTM.

File: src/github.com/getlantern/flashlight/config/config.go (hunk 6 above)
Comment (myleshorton) [inline src/github.com/getlantern/flashlight/config/config.go]: That from when I started using the chained URL in the config, so it was necessary @oxtoacart. I'll change it back now that there's no chained URL in the config though.

**Comment by myleshorton** [comment]
I changed StartPolling back to not be a method of Config @oxtoacart, and I already fixed the extra else.
